src/main/java/com/iota/iri/service/TipsManager.java

```diff
@@ -232,17 +259,20 @@ Hash randomWalk(final Set<Hash> visitedHashes, final Map<Hash, Long> diff, final
             } else {
                 // walk to the next approver
                 tips = tipSet.toArray(new Hash[tipSet.size()]);
-                if (!ratings.containsKey(tip)) {
-                    serialUpdateRatings(myApprovedHashes, tip, ratings, analyzedTips, extraTip);
+                if (!cumulativeWeights.containsKey(tip.getSubHash())) {
+                    cumulativeWeights.putAll(calculateCumulativeWeight(myApprovedHashes, tip, extraTip != null,
+                            analyzedTips));
                     analyzedTips.clear();
                 }
 
                 walkRatings = new double[tips.length];
                 double maxRating = 0;
-                long tipRating = ratings.get(tip);
+                ByteBuffer subHash = tip.getSubHash();
+                long tipRating = cumulativeWeights.get(subHash);
                 for (int i = 0; i < tips.length; i++) {
+                    subHash = tips[i].getSubHash();
                     //transition probability = ((Hx-Hy)^-3)/maxRating
-                    walkRatings[i] = Math.pow(tipRating - ratings.getOrDefault(tips[i],0L), -3);
+                    walkRatings[i] = Math.pow(tipRating - cumulativeWeights.getOrDefault(subHash,0), -3);
```

Comment:
Performance could be improved by a factor of 20+. See https://github.com/iotaledger/iri/pull/535

Reply:
I recommend leaving this change out of this PR, it's big enough as it is. It's an unrelated issue.

Reply:
We are supposed to switch to the formula described in the whitepaper (with the alpha). If it won't happen soon then we will merge your PR.